core/common/src/main/java/alluxio/master/MasterInquireClient.java

```diff
@@ -40,6 +43,81 @@
    */
   List<InetSocketAddress> getMasterRpcAddresses() throws UnavailableException;
 
+  /**
+   * Returns a canonical connect string representing how this client connects to the master.
+   *
+   * @return the connect string
+   */
+  ConnectString getConnectString();
+
+  /**
+   * Class for representing master inquire connect strings.
+   *
+   * Canonical connect strings are unique so that if two inquire clients have the same
+   * connect string, they connect to the same cluster.
+   *
+   * Examples: "zk://host1:port1,host2:port2/leader/path", "master:19998"
+   */
+  class ConnectString {
```

Comment:
Could this just be a string? Each implementation of `MasterInquireClient` should implement the logic to create the string, not call a method from the `ConnectString` class (ie. if I were to implement another inquire client, I shouldn't need to modify the interface class)?

Reply:
1) **Storing connect string construction logic in `MasterInquireClient` subclasses.**
The issue is that in `AbstractFileSystem` we have some configuration and want to know whether the connect string it implies is the same as the current `MasterInquireClient`. It feels weird to construct a client just to compare its connect string and then immediately discard it. If we construct a client for connect string comparison, we have to worry about whether it needs to be closed, and whether anything expensive happens in the constructor. I can also imagine that a new `MasterInquireClient` might actually be able to reuse `singleMasterConnectString` or `multiMasterConnectString`.
2) **Using String instead of ConnectString**
We could use String, but then we lose some type safety and expressiveness. As it is now, all instances of `ConnectString` are guaranteed to be in one of the 3 formats we define in `MasterInquireClient`. The advantage I see for using `String` is that we save code, but I think its worth it to give extra meaning and type safety to `ConnectString` instances. No-one can accidentally format a connect string in the wrong way.